pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/QueryStep.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.functions.transforms;
+
+import com.datastax.oss.pulsar.functions.transforms.datasource.QueryStepDataSource;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericArray;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.common.schema.SchemaType;
+
+/**
+ * Compute AI Embeddings for one or more records fields and put the value into a new or existing
+ * field.
+ */
+@Builder
+@Slf4j
+public class QueryStep implements TransformStep {
+
+  @Builder.Default private final List<String> fields = new ArrayList<>();
+  @Builder.Default private final String outputFieldName;
+  @Builder.Default private final String query;
+  @Builder.Default private final QueryStepDataSource dataSource;
+  private final Map<Schema, Schema> avroValueSchemaCache = new ConcurrentHashMap<>();
+
+  @Override
+  public void process(TransformContext transformContext) {
+    if (!isSchemaCompatible(transformContext)) {
+      return;
+    }
+
+    Map<String, Object> values = new HashMap<>();
+    collectFieldValuesFromKey(fields, transformContext, values);
+    collectFieldValuesFromValue(fields, transformContext, values);
+    List<Object> params = fields.stream().map(values::get).collect(Collectors.toList());
+    final List<Map<String, String>> results = dataSource.fetchData(query, params);
+    applyResultsToRecord(transformContext, results);
+  }
+
+  private void applyResultsToRecord(
+      TransformContext transformContext, List<Map<String, String>> results) {
+    final SchemaType type = transformContext.getValueSchema().getSchemaInfo().getType();
+    switch (type) {
+      case AVRO:
+        applyResultsToAvroRecord(transformContext, results);
+        break;
+      case JSON:
+        applyResultsToJsonRecord(transformContext, results);
+        break;
+      default:
+        return;
+    }
+  }
+
+  private void applyResultsToAvroRecord(
```

Comment:
Should use `TransformContext::addOrReplaceAvroValueFields`

Reply:
I agree, we should refactor and use common methods, I would do it in a follow up step